src/Npgsql.SourceGenerators/GeneratedDatabaseInfo.snbtxt

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Npgsql.Internal;
+using Npgsql.PostgresTypes;
+using Npgsql.Util;
+
+#nullable enable
+
+namespace Npgsql;
+
+partial class GeneratedDatabaseInfo
+{
+    static readonly Lazy<PostgresType[]> Types = new(() =>
+    {
+        ////////////////
+        // Base types //
+        ////////////////
+        {{ for base_type in base_types }}var t_{{ base_type.name }} = new PostgresBaseType("pg_catalog", "{{ base_type.name }}", {{ base_type.oid }}u);
+        {{ end }}
+
+        //////////////////
+        // Pseudo types //
+        //////////////////
+        {{ for pseudo_type in pseudo_types }}var t_{{ pseudo_type.name }} = new PostgresBaseType("pg_catalog", "{{ pseudo_type.name }}", {{ pseudo_type.oid }}u);
+        {{ end }}
+
+        /////////////////
+        // Range types //
+        /////////////////
+        {{ for range_type in range_types }}var t_{{ range_type.name }} = new PostgresRangeType("pg_catalog", "{{ range_type.name }}", {{ range_type.oid }}u, t_{{ range_type.element_name }});
+        {{ end }}
+
+        //////////////////////
+        // Multirange types //
+        //////////////////////
+        {{ for multirange_type in multirange_types }}var t_{{ multirange_type.name }} = new PostgresMultirangeType("pg_catalog", "{{ multirange_type.name }}", {{ multirange_type.oid }}u, t_{{ multirange_type.element_name }});
+        {{ end }}
+
+        /////////////////
+        // Array types //
+        /////////////////
+        {{ for array_type in array_types }}var t_{{ array_type.name }} = new PostgresArrayType("pg_catalog", "{{ array_type.name }}", {{ array_type.oid }}u, t_{{ array_type.element_name }});
+        {{ end }}
+
+        return new PostgresType[]
+        {
+            // Base types
+            {{ for base_type in base_types }}t_{{ base_type.name }},{{ end }}
```

Comment:
nit: I'd personally just inline all the instantiations directly into the array initializations, rather than have the temporary variables above; this would mean we generate a really trivial static readonly array initialized in-place, similar to PostgresMinimalDatabaseInfo.Types (but without reflection).

Reply:
> I'd personally just inline all the instantiations directly into the array initializations

This doesn't work for `PostgresRangeType`, `PostgresMultirangeType` and `PostgresArrayType` since they need instances of their base types for the constructor.

We could inline the instantiations of `PostgresArrayType` but that would rather complicate things for little benefit.

Reply:
Ok, I've amended the code to just create variables where it is necessary. It is a tiny bit more complicated now, but this should still be ok.

Reply:
Right, makes sense - I forgot about that part. No need to complicate anything, it's just a nit about how the generated code looks (nobody cares much).